### DISCUSSION TITLE
Add spell-checking to lint process

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#171] [Ideas]
Possibly grammar-checking too?

I've never looked into adding such a check before, so I'm not sure what solutions are out there.

If anyone decides to research this, please share your findings here. Please comment here before doing any work on a PR, so we can discuss. Thanks!

**Comment by MarlonPassos-git** [comment]
Interesting point. I usually use the [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) extension to help with that. Now, to execute this programmatically is another matter.

**Reply by aleclarson** [reply]
Oooh, it looks like that runs on CSpell, so @adamhamlin's upcoming PR work can be reused by it. We should recommend this extension [here](https://github.com/radashi-org/radashi/blob/main/.vscode/extensions.json) and [here](https://github.com/radashi-org/radashi/blob/main/.devcontainer/devcontainer.json#L18-L22) once we get CSpell configured and merged. 👍

**Comment by adamhamlin** [comment]
[eslint-plugin-spellcheck](https://www.npmjs.com/package/eslint-plugin-spellcheck) seems pretty standard. Tho wouldn't cover grammar.

@aleclarson is there a reason we don't have a "generalized" eslint setup? I see [this script](https://github.com/radashi-org/radashi/blob/c5057757803cb217001c57240f53f920a666057b/scripts/lint-compat.sh#L1) that specifically checks browser compatibility rules, why not just check-in an eslint config file?

**Reply by aleclarson** [reply]
> is there a reason we don't have a "generalized" eslint setup? I see [this script](https://github.com/radashi-org/radashi/blob/c5057757803cb217001c57240f53f920a666057b/scripts/lint-compat.sh#L1) that specifically checks browser compatibility rules, why not just check-in an eslint config file?

Just a case of trying to be too clever and failing. 😅  I will definitely change that..

**Comment by aleclarson** [comment]
[CSpell](https://cspell.org/) also looks like a valid option. (It has an [ESLint plugin](https://www.npmjs.com/package/@cspell/eslint-plugin) and [a CLI](https://www.npmjs.com/package/cspell)).

Note that I'd like to spell-check both source code and the docs, so that's something to keep in mind.

**Reply by adamhamlin** [reply]
Yeah CSpell looks pretty sweet, few more bells and whistles and can obviously handle more file types (e.g., the docs).

I'd probably suggest the following:
- Normalize the eslint setup, replacing the npm target `lint:compat` with `lint:eslint` or similar
- Include the `@cspell/eslint-plugin` in the eslint config, which itself points to a cspell config file
- Create a new npm target like `lint:docs` which uses the `cspell` cli on `*.mdx` files and points to the same cspell config file above

We could obviously just use the cspell cli for all files that we need, but I think we're going to want general eslinting at some point anyway and there's a real performance benefit to only having to parse the AST once as part of the lint process.

@aleclarson if you're amenable I can take this on

**Reply by aleclarson** [reply]
I'm in the midst of switching the ESLint setup, but I'd be happy to let you take on the CSpell integration once I'm done 👍 

> * Normalize the eslint setup, replacing the npm target `lint:compat` with `lint:eslint` or similar
> 
> * Include the `@cspell/eslint-plugin` in the eslint config, which itself points to a cspell config file
> 
> * Create a new npm target like `lint:docs` which uses the `cspell` cli on `*.mdx` files and points to the same cspell config file above

Sounds like a plan!

**Reply by adamhamlin** [reply]
Cool, lemme know

**Reply by aleclarson** [reply]
> I think we're going to want general eslinting at some point anyway

In case you didn't know, we're already using [Biome](https://biomejs.dev/) as a replacement for ESLint in terms of catching mistakes and a couple basic formatting rules (it replaces Prettier entirely for the source code). Unfortunately, they don't have a plugin system *yet*, so we're in this awkward transition period where we have two linters in the project.

**Reply by aleclarson** [reply]
Hey @adamhamlin, so I've switched off of ESLint entirely [in this commit](https://github.com/radashi-org/radashi/commit/1adaa4ede358803cc37a000f0585c6b2c6337ec1), because `eslint-plugin-compat` turned out to be pretty lackluster (it doesn't check method use like `array.flat()`). Therefore, we should use CSpell's CLI instead of the ESLint plugin.

That said, I'm ready for your PR 👍

**Reply by adamhamlin** [reply]
>In case you didn't know, we're already using [Biome](https://biomejs.dev/) as a replacement for ESLint in terms of catching mistakes and a couple basic formatting rules

Ahh right, I definitely saw that at some point and forgot, esp. the other day when I noticed we were using eslint (and prettier) somewhere else.

But yeah all the rest of that sounds good. Will use the `cspell` binary for any/all spellchecking we want to do 👍

**Reply by aleclarson** [reply]
Quick update: I refactored the `lint` command into TypeScript. The need came up when I added the `./scripts/` folder to be linted, which resulted in missing dependency errors since `./scripts/*/package.json` dependencies are only installed on-demand. Anyway, you'll want to add the spell-checking command to [this array](https://github.com/radashi-org/radashi/blob/main/scripts/lint/lint.ts#L11). (As a bonus, I got to remove [concurrently](https://github.com/open-cli-tools/concurrently) as a dependency, opting to write my own, which turned out to be not *that* complicated, so now we don't have Lodash in our lockfile anymore 🤣).

**Reply by adamhamlin** [reply]
Sweet. And yep, I briefly started working on this last night and that array is where I put it :)
